ports/rp2/mpbthciport.c

```diff
@@ -50,46 +53,60 @@ void mp_bluetooth_hci_init(void) {
     events_task_is_scheduled = false;
 }
 
-STATIC void mp_bluetooth_hci_start_polling(void) {
-    events_task_is_scheduled = false;
-    mp_bluetooth_hci_poll_now();
-}
-
 static int64_t mp_bluetooth_hci_timer_callback(alarm_id_t id, void *user_data) {
     poll_timer_id = 0;
-    mp_bluetooth_hci_poll_now();
+    mp_bluetooth_hci_schedule_poll();
     return 0;
 }
 
+// The bt stack uses this for command timeouts, so it shouldn't be used elsewhere
 void mp_bluetooth_hci_poll_in_ms(uint32_t ms) {
+    debug_printf("mp_bluetooth_hci_poll_in_ms %lu\n", ms);
+    if (poll_timer_id > 0) {
+        cancel_alarm(poll_timer_id);
+    }
     poll_timer_id = add_alarm_in_ms(ms, mp_bluetooth_hci_timer_callback, NULL, true);
 }
 
 // For synchronous mode, we run all BLE stack code inside a scheduled task.
-// This task is scheduled periodically via a timer, or immediately after UART RX IRQ.
-STATIC mp_obj_t run_events_scheduled_task(mp_obj_t none_in) {
-    (void)none_in;
+STATIC mp_obj_t run_events_scheduled_task(mp_obj_t obj) {
     events_task_is_scheduled = false;
-    // This will process all buffered HCI UART data, and run any callouts or events.
+    // This will process all HCI data, and run any callouts or events.
     mp_bluetooth_hci_poll();
     return mp_const_none;
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(run_events_scheduled_task_obj, run_events_scheduled_task);
 
-// Called periodically (systick) or directly (e.g. UART RX IRQ) in order to
-// request that processing happens ASAP in the scheduler.
-void mp_bluetooth_hci_poll_now(void) {
+// Perform processing ASAP in the scheduler.
+void mp_bluetooth_hci_schedule_poll(void) {
     if (!events_task_is_scheduled) {
-        events_task_is_scheduled = mp_sched_schedule(MP_OBJ_FROM_PTR(&run_events_scheduled_task_obj), mp_const_none);
+        events_task_is_scheduled = mp_sched_schedule(MP_OBJ_FROM_PTR(&run_events_scheduled_task_obj), NULL);
         if (!events_task_is_scheduled) {
             // The schedule queue is full, set callback to try again soon.
             mp_bluetooth_hci_poll_in_ms(5);
         }
     }
 }
 
+int mp_bluetooth_hci_uart_deinit(void) {
+    debug_printf("mp_bluetooth_hci_uart_deinit\n");
+
+    // If a poll callback is set cancel it now.
+    if (poll_timer_id > 0) {
+        cancel_alarm(poll_timer_id);
+    }
+    poll_timer_id = 0;
+    return 0;
+}
+
+#if !MICROPY_PY_BLUETOOTH_CYW43
```

Comment:
Better to not use negation and change the logic to `#if MICROPY_PY_BLUETOOTH_CYW43`.

Reply:
- Tricky. It looks like cyw43 only needs a couple of functions in here. It doesn't need any of the functions that deal with uarts. For the time being I've wrapped this in a locally defined USES_UARTS macro